### ISSUE TITLE
[ENH] Add classical Online Portfolio Selection algorithms

### ISSUE THREAD

**Issue by lcrmorin** [skfolio/skfolio#261]
`skfolio.optimization` covers convex allocation (`MeanRisk`, `RiskBudgeting`, `DistributionallyRobustCVaR`...), hierarchical clustering (`HierarchicalRiskParity`, `HERC`, `NestedClusteredOptimization`, `SchurComplementary`), ensembling (`StackingOptimization`) and naive baselines. Every one of these estimates weights from a  `ReturnDistribution` (expected returns + covariance) supplied by a `prior_estimator`.

There's no support for the classical "Online Portfolio Selection" (OPS) literature, Cover's Universal Portfolio, multiplicative-update and mean-reversion strategies, where weights update sequentially, period-by-period, directly from realized price relatives, typically without ever estimating a covariance matrix, and carry worst-case regret guarantees relative to the best fixed-weight portfolio in hindsight. `skfolio.model_selection._online` already ships an evaluation harness (`OnlineGridSearch`, `online_predict`, `online_score`) for periodically-refit estimators, but nothing built for genuinely sequential/multiplicative weight updates sits underneath it.

**Describe the solution you'd like**

A small family of estimators under `skfolio.optimization.online`, each a `BaseOptimization` subclass consistent with the existing `weights_` contract; Including at least:

- `ConstantRebalancedPortfolio` (Best-CRP baseline)
- `ExponentiatedGradient` (Helmbold, Schapire, Singer & Warmuth, 1998)
- `OLMAR` (Li & Hoi, 2012)
- `PAMR` (Li, Zhao, Hoi & Gopalkrishnan, 2012)

Each processes the full returns matrix `X` at `fit` and iterates period-by-period internally (the same "whole-sample loop inside `fit`" pattern `HierarchicalRiskParity` already uses), producing a final `weights_` vector and optionally the full trajectory (e.g. `weights_path_`) for inspection.

**Describe alternatives you've considered**

Repeatedly refitting an existing convex optimizer via `WalkForward` instead. Not ideal. 

**Additional context**

References: 
Cover (1991) "Universal Portfolios"; 
Helmbold, Schapire, Singer & Warmuth (1998) "On-Line Portfolio Selection Using Multiplicative Updates"
Li & Hoi (2012) "On-Line Portfolio Selection with Moving Average Reversion"
Li, Zhao, Hoi & Gopalkrishnan (2012) "PAMR: Passive Aggressive Mean Reversion Strategy for Portfolio Selection"
Li & Hoi (2014) "Online Portfolio Selection: A Survey" (ACM Computing Surveys) as the main reference.

Happy to open a PR - if this direction looks right.

**Comment by CarloNicolini** [comment]
Please look at my fork.
I’ve done really extensive work on that and deeply thought about it with some small deviations from skfolio API.
It’s exactly what you described, happy to see some interest!

https://github.com/CarloNicolini/skfolio/tree/universal_portfolio

Look at the examples and documentation for an overview.
I’ve not only implemented the classics OLPS method from Li and Hoi, but I’ve interpreted them under a larger umbrella based on recent online optimization point of view. I took inspiration from textbooks like Elad Hazan and Francesco Orabona. Methods can be composed easily and they all have extensive amount of parameters like constraints etc in the same style as the MeanRisk estimator.
Since it was a rather long and complex job, it still may have some rough edges, but nonetheless I think it’s pretty rich and comprehensive.
Interestingly I’ve managed to cover multiple risk measures and objectives, not only strongly convex cases. It also has a part about regret computation under various settings.

Maybe it’s time to think about a strategy to make a PR.

**Comment by lcrmorin** [comment]
Very interesting stuff. I thin there is a demand. If you are okay I'll start to look into your code for evaluating the remaining work for a PR (if any).

**Comment by CarloNicolini** [comment]
For the sake of cleanness, having autograd in the dependencies is no longer necessary as I've used it to verify the correctness of the analytical gradients w.r.t. asset weights of various risk metrics reimplementing them in terms of 'autograd.numpy as np' and even numerical differentiation via finite differences.

Analytical versions of gradients are clearly faster to run and are used by downstream estimators internally.

**Comment by lcrmorin** [comment]
Dang you read my mind, I was drafting some comments regarding the need to use autograd as a dependency. 

Other topics:
- I count nearly 20k lines and a version behind main, this mean rebasing and multiples PRs
- there are datasets (1.5 Mb), is this something we want to keep ?

**Comment by CarloNicolini** [comment]
The large number of lines is an extensive set of automated pytests covering many stuff. 
The online module is around 10k lines in total.
The version was behind `main' by 99 commits I just saw but I've just synced it to main without problems.

About the datasets, I agree they could stay outside. I just wanted to replicate some of the Li and Hoi results so I've used their very same datasets (which I've gathered here and there in the web) and from the Matlab OLPS toolbox.

They are clearly not necessary for a PR. 
I let @HugoDelatte decide about them.

I should gather a document where I've started writing the technical approach I've used for these online methods.

**Comment by HugoDelatte** [comment]
Yes agree on the datasets, we should reuse the existing skfolio ones if possible.

@CarloNicolini did a very impressive amount of work on that branch, both on the algorithms themselves and on the supporting abstractions needed to make them usable in skfolio at the time.

We intentionally waited for the stable v1 release before integrating a feature of this scale because we first needed to settle the broader online learning API across skfolio.

Since Carlo's branch was implemented, a lot of the surrounding infrastructure has landed in skfolio, including incremental moment and prior estimators, online `MeanRisk`, characteristic factor models, `partial_fit`, `online_predict`, and online hyperparameter tuning.

Several responsibilities that Carlo had to implement locally can therefore now be delegated to common skfolio abstractions. This creates some API divergence between the branch and v1, but the underlying work remains highly reusable.

I think the main challenge now is agreeing on how to integrate it cleanly with the APIs that have since been established. Given the scale and novelty of the feature, I think it is worth spending some time discussing the architecture here before starting the PR.

I'll start with some high level ideas:

* I think we should reuse as much of the algorithmic work from Carlo's branch as possible, while adapting the surrounding API to the online infrastructure that now exists in skfolio.

  The end goal should be that online portfolio estimators behave like regular skfolio estimators and work directly with the existing online model selection tools.

  For example:

  ```python
  model = ExponentiatedGradient(
      learning_rate=0.05,
      max_turnover=0.10,
  )

  pred = online_predict(
      model,
      X,
      warmup_size=252,
      test_size=1,
  )
  ```

  and similarly:

  ```python
  search = OnlineGridSearch(
      ExponentiatedGradient(),
      param_grid={
          "learning_rate": [0.01, 0.03, 0.1],
      },
      warmup_size=252,
      test_size=1,
  )
  ```

  Carlo's branch implements its own online prediction machinery because the generic model selection layer did not exist at the time. In v1 I think this responsibility can now move to `online_predict`, `online_score`, `OnlineGridSearch`, and `OnlineRandomizedSearch`.

* I think we need a `BaseOnlineOptimization(BaseOptimization)` following the same lifecycle as the rest of the online API.

  The contract would be:

  ```python
  fit(X)
  # reset the state and process X causally

  partial_fit(X)
  # continue from the current state
  ```

  Both should accept either one observation or a block of observations, but internally they should always process observations sequentially.

  Something like:

  ```python
  def fit(self, X, y=None, **fit_params):
      self._reset()
      return self.partial_fit(X, y, **fit_params)

  def partial_fit(self, X, y=None, **fit_params):
      for X_t, y_t, params_t in self._iter_observations(
          X, y, fit_params
      ):
          self._update_one(X_t, y_t, **params_t)

      return self
  ```

  The row by row processing is important because `online_predict` can pass a complete warmup block to `partial_fit`. We still need the internal updates to remain causal.

  This also means that Carlo's `warm_start` can likely be replaced by the standard v1 convention where `fit` resets and `partial_fit` continues.

* We should follow the v1 as-of indexing convention explicitly.

  After processing observation `t`, `weights_` should represent the target allocation as of `t`, computed using only information available through `t`, and intended for the next return period.

  Conceptually:

  ```text
  returns_t
  metadata_t
      |
      v
  partial_fit
      |
      v
  prior state as of t
  algorithm state as of t
  weights_ as of t
      |
      v
  used for return t + 1
  ```

  This also gives a clear meaning to fitted attributes such as:

  ```python
  initial_weights_
  # allocation before the first observation

  weights_
  # allocation as of the latest processed observation
  ```

  I think it is important that the documentation explicitly explains the index shift with the notation commonly used in the online portfolio literature, where the same update is often written as `b[t + 1] = U(b[t], x[t])`.

* `prior_estimator` should probably become a first class optional parameter.

  This is one of the areas where v1 gives us infrastructure that did not exist when the branch was originally implemented.

  For example:

  ```python
  model = ExponentiatedGradient(
      learning_rate=0.05,
      prior_estimator=EmpiricalPrior(
          mu_estimator=EWMu(...),
          covariance_estimator=EWCovariance(...),
      ),
      max_variance=0.02**2,
  )
  ```

  or:

  ```python
  model = ExponentiatedGradient(
      prior_estimator=CharacteristicsFactorModel(...),
      linear_constraints=[
          "Momentum >= 0.10",
          "Quality >= 0.05",
      ],
  )
  ```

  The classical online algorithm can still remain distribution free. I think `prior_estimator=None` should therefore really mean no prior, unlike `MeanRisk` where a prior is fundamental to the optimization.

  When a prior is provided, it should also be updated causally at each observation:

  ```python
  self.prior_estimator_.partial_fit(
      X_t,
      y_t,
      **params_t,
  )

  return_distribution = (
      self.prior_estimator_.return_distribution_
  )
  ```

  The resulting `ReturnDistribution` can then be used by risk constraints, factor constraints, stressed scenarios, uncertainty sets, and custom constraints.

  For online usage I think it is reasonable to require the supplied prior and its children to support `partial_fit`. I would avoid silently refitting a batch-only prior on the complete history after every observation.

* I think we should support the full skfolio constraint vocabulary and reuse the same parameter names and semantics as `MeanRisk`.

  For example:

  ```python
  model = OLMAR(
      window=5,
      eps=10,
      prior_estimator=...,
      min_weights=0,
      max_weights=0.10,
      budget=1,
      groups=groups,
      linear_constraints=[
          "Technology <= 0.30",
          "Momentum >= 0.10",
      ],
      max_turnover=0.10,
      max_variance=0.02**2,
  )
  ```

  This would be one of the main advantages of integrating these algorithms deeply into skfolio rather than exposing only their textbook simplex versions.

  I would prefer adapting Carlo's parameters to the existing v1 vocabulary where possible. For example a raw `covariance` input can now be represented by `prior_estimator`, and `variance_bound` can align with `max_variance`.

  Architecturally, I think the common constraint and risk expression machinery should ideally be extracted from `ConvexOptimization` into reusable internals.

  Conceptually:

  ```text
                   shared constraint machinery
                    /                       \
                   /                         \
          ConvexOptimization          BaseOnlineOptimization
  ```

  The online estimators would then consume the same implementation for bounds, groups, factor constraints, turnover, cardinality, risk limits, and custom constraints without having to inherit from `ConvexOptimization`.

* One area where I think the v1 architecture lets us generalize Carlo's approach is the interaction between the online update and the constraints.

  Carlo's generic projection machinery is a very useful abstraction and gives us a strong starting point. With the richer constraint infrastructure now available in v1, I think we can preserve the native geometry of each algorithm when solving the constrained update.

  For example, Exponentiated Gradient naturally uses entropy or KL geometry while algorithms such as OLMAR and PAMR naturally use Euclidean geometry.

  Instead of always doing:

  ```python
  candidate = self._raw_update(...)
  weights = l2_project(candidate, constraints)
  ```

  the common abstraction could be closer to:

  ```python
  weights = self._solve_update(
      current_weights=self.weights_,
      observation=X_t,
      return_distribution=return_distribution,
      constraints=constraints,
  )
  ```

  For simple cases we can keep the analytical formulas.

  For richer constraints we can formulate the equivalent optimization problem with CVXPY. This would let us support the full skfolio constraint set while staying as close as possible to the original update rule of each algorithm.

* I think concrete public estimators would integrate better with the current sklearn style API than a public `FollowTheWinner(strategy=...)` or `FollowTheLoser(strategy=...)` interface.

  For example:

  ```python
  ExponentiatedGradient(...)
  OnlineGradientDescent(...)
  OLMAR(...)
  PAMR(...)
  CWMR(...)
  RMR(...)
  ```

  This gives cleaner parameter spaces and makes hyperparameter tuning straightforward:

  ```python
  OnlineGridSearch(
      OLMAR(),
      {
          "window": [3, 5, 10],
          "eps": [5, 10, 20],
          "max_turnover": [0.05, 0.10],
      },
  )
  ```

  Carlo's `FollowTheWinner` and `FollowTheLoser` abstractions are still very useful internally because they expose the common structure between a large family of algorithms. The OCO machinery, mirror maps, predictors, and update kernels can remain shared implementation underneath concrete public estimators.

  We may also want to revisit some of the public class names. skfolio generally tries to avoid acronyms when a descriptive name makes the API easier to discover and understand for users who are not already familiar with a specific paper or literature.

  For example, instead of exposing only:

  ```python
  OLMAR(...)
  PAMR(...)
  CWMR(...)
  ```

  we could consider more descriptive names, while keeping the literature name and acronym clearly documented.

  Something along the lines of:

  ```python
  OnlineMovingAverageReversion(...)
  PassiveAggressiveMeanReversion(...)
  ConfidenceWeightedMeanReversion(...)
  ```

  The exact naming should be discussed because some expanded names can also become too long, but I think API readability is worth considering before we freeze the public classes.

* Weight drift and exact transaction cost execution need special care.

  Carlo's branch currently handles drift, wealth, and rebalance accounting directly because there was no common skfolio execution abstraction available at the time.

  With #161, I think we now have the opportunity to make this common infrastructure shared by online optimizers, `MeanRisk`, and any other strategy that needs path dependent execution.

  The separation I have in mind is:

  ```text
  optimizer
      |
      + target weights
      + constraints
      + learning state

  portfolio / model selection
      |
      + weight drift
      + pre-trade holdings
      + realized trades
      + exact transaction costs
      + compounded wealth
  ```

  This is particularly important for `max_turnover`. We need one consistent definition of whether turnover is measured relative to the previous target weights or the drifted pre-trade holdings.

  I think #161 and this feature can be developed partly in parallel, but we should settle that semantic boundary before finalizing the online optimization API.

* There is also a small point to clarify around `online_predict`.

  These algorithms are normally updated after each observation. With `test_size > 1`, `online_predict` holds one allocation across several observations before giving that block back to `partial_fit`.

  If `partial_fit` then updates the strategy row by row, it would create intermediate allocations that were never actually used.

  I think we should either restrict these estimators to `test_size=1` in `online_predict`, or define very clearly what a larger `test_size` means for them.

* Finally, several responsibilities currently handled directly inside Carlo's branch can likely be delegated to common v1 abstractions.

  This includes `predict_online`, custom `fit_predict`, `warm_start`, wealth bookkeeping, raw covariance handling, and tracking data stored directly on the estimator.

  These abstractions were necessary to make the branch work end to end when it was developed. Today we can probably simplify the online optimization layer by letting the estimator focus on the causal learning problem:

  ```python
  model.partial_fit(X_t)

  model.weights_
  model.prior_estimator_
  ```

  The model selection layer can handle out of sample prediction:

  ```python
  online_predict(model, X)
  ```

  and the portfolio layer can handle evaluation and eventually the execution semantics from #161.

Overall, I think Carlo's branch gives us most of the difficult algorithmic machinery already. The main work now is to adapt the surrounding abstractions so the feature feels native to the v1 API and composes cleanly with the online functionality that has since been added.

**Comment by lcrmorin** [comment]
Thanks for this message. I'll take a bit of time to digest it.

**Comment by lcrmorin** [comment]
Thanks for the detailed writeup, I really appreciate the thought put into this. Without having delved too much in Carlo's code myself, I think we can reuse Carlo's algorithmic core (update rules, mirror maps, OCO machinery).

I'd say there are two open questions:

- @CarloNicolini this is a bigger ask than "split the branch into smaller PRs,". It's re-hosting the algorithms inside a new abstraction (BaseOnlineOptimization, prior_estimator as first-class, constraint machinery shared with ConvexOptimization, geometry-native updates.... ). Would you rather lead that port yourself, or would it help if the plumbing (base class + extracted constraint machinery) got scaffolded first?
- @HugoDelatte the turnover/TC semantics point to #161. Saw #286 is already in review and settles the semantics (turnover measured against held/drifted weights, not previous targets). Is there help needed on this?

**Comment by CarloNicolini** [comment]
Since I still did not have time to fully deep dive in the 1.0 API I'd prefer to let you guys discuss here what the best scaffold could be where the OCO algorithms can go.
Since I had put lot of effort in replicating the pre 1.0 MeanRisk constraints machinery I don't think it should be too complicated to port those constraints to the current OnlinePortfolioSelection base class, as one has to change some internal helper methods.

However from the point of view of redefining some internals of the methods for the sake of reusing the new weights drift machinery I believe we have to make a rather detailed plan first.
Prequential-sequential estimators and maintaining the right order of operations in the OCO methods was very important to avoid leakage of the future in the present, and accumulating the previous weights at the right time was paramount.
So yes, some algorithmic brain squeeze is required that is not just scaffolding or wrapping stuff.

By the way, some methods like Adagrad (if my memory is not mistaking me) make use of internal covariance matrices that are not related to returns covariance but to internal numerical states like gradients etc...thus it should not make sense to provide estimators for them.

**Comment by lcrmorin** [comment]
So from my understanding (or for my understanding), we need to keep track three kinds of states: 

1. **Algorithm state:** previous targets, gradients, learning rates, and adaptive matrices.  These stares are kept by the optimizer.
2. **Statistical state:** return covariance and other estimates. They are supplied by an optional prior estimator.
3. **Execution state:** actual holdings, weight drift, trades, and realised wealth. This is handled by the portfolio, and goes trough evaluation infrastructure.

In particular, for the general problem at hand, the gradient based matrices are different from prior covariance estimates and should go in algorithm state...

For an Exponentiated Gradient MVP, I drafted the approach:

- Build on the weight-drift infrastructure merged in #286, while checking the update ordering carefully.
- Start without a prior estimator, since basic EG does not require covariance estimates.
- Have `fit(X)` reset state and `partial_fit(X)` continue, accepting either one row or a block processed causally.
- After observing period *t*, let `weights_` represent the target for *t+1*, with an explicit initial allocation.
- Calculate turnover from actual drifted holdings at rebalance time, keeping any previous-target state needed by the algorithm separate.
- Leave constraints out for the moment. 
- Initially use `test_size=1`, leaving larger-window semantics for a separate discussion.

Would this be a reasonable first scope? Do you see any issue with the approach? @CarloNicolini @HugoDelatte
